compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -675,19 +700,148 @@ pub enum OrphanCheckErr<'tcx> {
 /// 1. All type parameters in `Self` must be "covered" by some local type constructor.
 /// 2. Some local type must appear in `Self`.
 #[instrument(level = "debug", skip(tcx), ret)]
-pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanCheckErr<'_>> {
-    // We only except this routine to be invoked on implementations
+pub fn orphan_check<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_def_id: DefId,
+) -> OrphanCheckResult<'tcx, FxIndexMap<Symbol, Span>> {
+    // We only accept this routine to be invoked on implementations
     // of a trait, not inherent implementations.
-    let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap().instantiate_identity();
-    debug!(?trait_ref);
+    let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+    debug!(trait_ref = ?trait_ref.skip_binder());
 
     // If the *trait* is local to the crate, ok.
-    if trait_ref.def_id.is_local() {
-        debug!("trait {:?} is local to current crate", trait_ref.def_id);
-        return Ok(());
+    if let Some(def_id) = trait_ref.skip_binder().def_id.as_local() {
+        debug!("trait {def_id:?} is local to current crate");
+        return OrphanCheckResult::Ok;
+    }
+
+    let infcx = tcx.infer_ctxt().intercrate(true).build();
+    let cause = ObligationCause::dummy();
+    let args = infcx.fresh_args_for_item(cause.span, impl_def_id);
+    let trait_ref = trait_ref.instantiate(tcx, args);
+
+    match orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |user_ty| {
+        let ty::Alias(..) = user_ty.kind() else { return Ok(user_ty) };
+
+        let ocx = ObligationCtxt::new(&infcx);
+        let ty = ocx.normalize(&cause, ty::ParamEnv::empty(), user_ty);
+        let ty = infcx.resolve_vars_if_possible(ty);
+        let errors = ocx.select_where_possible();
+        if !errors.is_empty() {
+            return Ok(user_ty);
+        }
+
+        let ty = if infcx.next_trait_solver() {
+            let mut fulfill_cx = <dyn TraitEngine<'_>>::new(&infcx);
+            infcx
+                .at(&cause, ty::ParamEnv::empty())
+                .structurally_normalize(ty, &mut *fulfill_cx)
+                .map(|ty| infcx.resolve_vars_if_possible(ty))
+                .unwrap_or(ty)
+        } else {
+            ty
+        };
+
+        Ok(ty)
+    }) {
+        Ok(OrphanCheckResult::Ok) => OrphanCheckResult::Ok,
+        Ok(OrphanCheckResult::Issue99554(UncoveredTyParams { uncovered, local_ty })) => {
```

Comment:
merge w/ branch below

Reply:
Ah, that's a bit annoying to do. don't think I can merge this using an or-pattern since also need to bind it with `@`. might be doable but don't have the time to play around rn. Maybe later.

On second thought, I could replace my custom three-variant `Result`-like type with libcore's `Result` and just add a flag inside the `Err` payload type if the error is “strict” or “lenient” or whatever the terminology at the cost of ppl forgetting to look at said flag when modifying this code thereby introducing breakage, dunno.